### PR TITLE
Prevent tokenizing .pyc files in format_records

### PR DESCRIPTION
**Pull request by mfrodl**
This update should address issue https://github.com/joblib/joblib/issues/258.

### CODE REVIEW

**Comment by lesteve** [comment]
I had a look the other day and it seems that's what IPython.core.ultratb.VerboseTB is doing now so sounds good as far as I am concerned.

Is there an easy way to test this ?

**Comment by martin-frodl** [comment]
> Is there an easy way to test this ?

Is the reproducer given in #258 sufficient as a test?

**Comment by lesteve** [comment]
It looks like it could be the basis of a test but:
1. you would need to put it in the form of a test. Look at examples in joblib/test
2. I can't actually reproduce the behaviour on your simple example, i.e. I don't get the EOF messages but the full traceback on master.

File: joblib/format_stack.py [hunk 1]
```diff
@@ -196,6 +196,10 @@ def format_records(records):   # , print_globals=False):
             # keep the original file string.
             pass
         link = file
+
+        if file.endswith('.pyc'):
+            file = file[:-4] + '.py'
```
Comment (lesteve) [inline joblib/format_stack.py]: Can you put these two lines above the `link = file` line?

This way, the filename that appears in the traceback is a .py and not a .pyc.

Bonus points if you squash your commits.

**Comment by lesteve** [comment]
For the record I just tested and it looks like this fixes the original scikit-learn problem.

File: joblib/format_stack.py (hunk 1 above)
Comment (martin-frodl) [inline joblib/format_stack.py]: > Can you put these two lines above the link = file line?

That's where the added lines were meant to be, I must have accidentally swapped the places.

Also, I am still working on the test, so bear with me please. :)

Comment (lesteve) [inline joblib/format_stack.py]: > Also, I am still working on the test, so bear with me please. :)

No problem, just add a comment on this PR when you have it ready.

**Comment by martin-frodl** [comment]
I fixed the line order in the original commit and attached a test case as well. Please review.

**Comment by lesteve** [comment]
Your test look fine, thanks a lot ! Out of interest, did you manage to reproduce the original EOF problem ?

File: joblib/test/test_format_stack.py [hunk 2]
```diff
@@ -20,3 +21,23 @@ def __repr__(self):
...
+        records = _fixed_getframes(etb)
+
+        # Modify filenames in traceback records from .py to .pyc
+        pyc_records = []
+        for record in records:
+            if record[1].endswith('.py'):
+                record = list(record)
+                record[1] += 'c'
+                record = tuple(record)
+                pyc_records.append(record)
```
Comment (lesteve) [inline joblib/test/test_format_stack.py]: This is the kind of place list comprehensions shine IMHO:

```python
pyc_records = [(_1, filename + 'c' if filename.endswith('.py') else filename,
                _2, _3, _4, _5)
               for _1, filename, _2, _3, _4, _5 in records]
```

In this particular case maybe using a helper function would make it clearer.

**Comment by lesteve** [comment]
Merging this one thanks.

**Comment by GaelVaroquaux** [comment]
Thanks heap Martin!

**Pull request merged**